src/webgpu/util/fp_interval.ts

```diff
@@ -0,0 +1,238 @@
+import { assert } from '../../common/util/util.js';
+
+import { kValue } from './constants.js';
+import { flushSubnormalNumber, nextAfter, oneULP } from './math.js';
+
+export class FPInterval {
+  public readonly begin: number;
+  public readonly end: number;
+
+  public constructor(begin: number, end: number) {
+    assert(begin <= end, `begin (${begin}) must be equal or before end (${end})`);
+    assert(!Number.isNaN(begin), `begin must not be NaN`);
+    assert(!Number.isNaN(end), `end must not be NaN`);
+
+    if (begin === Number.NEGATIVE_INFINITY || begin < kValue.f32.negative.min) {
+      this.begin = Number.NEGATIVE_INFINITY;
+    } else if (begin === Number.POSITIVE_INFINITY || begin > kValue.f32.positive.max) {
+      this.begin = kValue.f32.positive.max;
+    } else {
+      this.begin = begin;
+    }
+
+    if (end === Number.POSITIVE_INFINITY || end > kValue.f32.positive.max) {
+      this.end = Number.POSITIVE_INFINITY;
+    } else if (end === Number.NEGATIVE_INFINITY || end < kValue.f32.negative.min) {
+      this.end = kValue.f32.negative.min;
+    } else {
+      this.end = end;
+    }
+  }
+
+  public contains(x: number): boolean {
+    return this.begin <= x && x <= this.end;
+  }
+
+  static span(...intervals: FPInterval[]): FPInterval {
+    assert(intervals.length > 0, `span of an empty list of FPIntervals is not allowed`);
+    const mins = new Set<number>();
+    const maxs = new Set<number>();
+    intervals.forEach(i => {
+      mins.add(i.begin);
+      maxs.add(i.end);
+    });
+    return new FPInterval(Math.min(...mins), Math.max(...maxs));
+  }
+
+  public toString(): string {
+    return `[${this.begin}, ${this.end}]`;
+  }
+}
+
+export interface FPUnaryIntervalBuilder {
+  singular(x: number): FPInterval;
+  range?(i: FPInterval): FPInterval;
```

Comment:
range is very intentionally optional, since it requires handling extremas in the interval correctly, and is only used if another function inherits accuracy from it. For monotonic functions implementing this is trivial, but needs custom code for non-monotonic functions, so I would only implement it for functions that we actually use it.